### PR TITLE
Add seekable csv source

### PR DESCRIPTION
**Pull request by srascar**
I am targetting this branch, because I want to implement a series of new feature that will
eventually cause BC-break.

## Changelog

```markdown
### Added
- Seekable csv source
```

## To do

- [x] Update the tests
- [ ] Update the documentation
- [ ] Add an upgrade note

## Subject

I want to add new features to exporter in order to improve it's functionalities.
The first step is to allow the source to be seekable.
This combined with a few other classes will provide much more flexibility and possibilities 
to the bundle

### CODE REVIEW

File: src/Source/SeekableSourceIteratorInterface.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\Exporter\Source;
+
+/**
+ * interface SeekableSourceIteratorInterface.
```
Comment (core23) [inline src/Source/SeekableSourceIteratorInterface.php]: Please remove this selfexplaining line or add a better description

File: src/Source/CsvSourceIterator.php [hunk 2]
```diff
@@ -145,13 +144,72 @@ public function rewind()
...
         }
 
         return true;
     }
+
+    /**
+     * Initialize read process setting CSV options
+     * and settings field names. (Assuming that first line is column name).
+     */
+    protected function initializeRead()
```
Comment (core23) [inline src/Source/CsvSourceIterator.php]: Can you make this `private` or at least `final`?

File: src/Source/CsvSourceIterator.php [hunk 3]
```diff
@@ -145,13 +144,72 @@ public function rewind()
...
+
+        if ($this->hasHeaders) {
+            $this->columns = $this->file->fgetcsv();
+        }
+    }
+
+    /**
+     * @return array|void
+     */
+    protected function getNextLine()
```
Comment (core23) [inline src/Source/CsvSourceIterator.php]: Can you make this `private` or at least `final`?

File: src/Source/CsvSourceIterator.php [hunk 4]
```diff
@@ -145,13 +144,72 @@ public function rewind()
...
+
+        return $this->combineColumns($line);
+    }
+
+    /**
+     * @param array $line
+     *
+     * @return array|void
+     */
+    protected function combineColumns($line)
```
Comment (core23) [inline src/Source/CsvSourceIterator.php]: Can you make this `private` or at least `final`?

**Comment by Soullivaneuh** [comment]
Changelog updated, please see and correct if I'm wrong.

File: src/Source/CsvSourceIterator.php [hunk 5]
```diff
@@ -14,17 +14,17 @@
 /**
  * Read data from a csv file.
  *
- * @author Vincent Touzet <vincent.touzet@gmail.com>
```
Comment (Soullivaneuh) [inline src/Source/CsvSourceIterator.php]: This should not be changed.

At least, you can add your author tag below, but not changing the current one.

**Comment by rande** [comment]
@sonata-project/contributors @srascar is going to send different PR to improve exporter, code will be BC with the current version, but will allow handling more use cases while working with exporting/importing data.

@srascar can you open an issue with the different steps of your work ? also can you add the missing part: documentation (and upgrade note ? => should not be required).

File: src/Source/CsvSourceIterator.php [hunk 6]
```diff
@@ -105,38 +106,37 @@ public function key()
...
-            $this->columns = $line;
-            $line = fgetcsv($this->file, 0, $this->delimiter, $this->enclosure, $this->escape);
+        $this->initializeRead();
+        $this->currentLine = $this->getNextLine();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function seek($position)
```
Comment (greg0ire) [inline src/Source/CsvSourceIterator.php]: Please make that method final

File: src/Source/CsvSourceIterator.php [hunk 7]
```diff
@@ -145,13 +145,72 @@ public function rewind()
...
         return true;
     }
+
+    /**
+     * Initialize read process setting CSV options
+     * and settings field names. (Assuming that first line is column name).
+     */
+    private function initializeRead()
+    {
+        $this->file = new \SplFileObject($this->filename);
```
Comment (rande) [inline src/Source/CsvSourceIterator.php]: You should close the file first in case of ``rewind`` ?

File: test/Source/CsvSourceIteratorTest.php [hunk 8]
```diff
@@ -90,4 +90,47 @@ public function testRewind()
...
+        $iterator = new CsvSourceIterator($this->filename, ',', '"', '\\', false);
+
+        $iterator->seek(1);
+        $value = $iterator->current();
+        $this->assertTrue(is_array($value));
+        $this->assertEquals('John 1', $value[0]);
+    }
+
+    /**
+     * @expectedException \RuntimeException
```
Comment (greg0ire) [inline test/Source/CsvSourceIteratorTest.php]: Please use `setExceptedException` instead

File: test/Source/CsvSourceIteratorTest.php [hunk 9]
```diff
@@ -90,4 +90,47 @@ public function testRewind()
...
+firstname,name
+John 1,Doe
+John 2,Doe, extra
+"John, 3", Doe
+EOF;
+        file_put_contents($this->filename, $csv);
+
+        $iterator = new CsvSourceIterator($this->filename);
+
+        foreach ($iterator as $value) {
```
Comment (greg0ire) [inline test/Source/CsvSourceIteratorTest.php]: What is this?

File: test/Source/CsvSourceIteratorTest.php (hunk 8 above)
Comment (srascar) [inline test/Source/CsvSourceIteratorTest.php]: I used `expectException` since `setExpectedException()` is now deprecated

**Comment by srascar** [comment]
@rande PR updated.

File: src/Source/CsvSourceIterator.php [hunk 10]
```diff
@@ -145,13 +145,77 @@ public function rewind()
...
+    /**
+     * Initialize read process setting CSV options
+     * and settings field names. (Assuming that first line is column name).
+     */
+    private function initializeRead()
+    {
+        // close the previous file
+        if ($this->file instanceof \SplFileObject) {
+            $this->file = null;
+        }
```
Comment (greg0ire) [inline src/Source/CsvSourceIterator.php]: If this closes the previous file just by loosing the last reference to it, how does line 165 not have the exact same effect?

File: src/Source/CsvSourceIterator.php [hunk 11]
```diff
@@ -145,13 +145,77 @@ public function rewind()
...
-
             return false;
         }
 
         return true;
     }
+
+    /**
+     * Initialize read process setting CSV options
+     * and settings field names. (Assuming that first line is column name).
```
Comment (greg0ire) [inline src/Source/CsvSourceIterator.php]: "settings" => "setting" ?

File: src/Source/CsvSourceIterator.php [hunk 12]
```diff
@@ -145,13 +145,77 @@ public function rewind()
...
+                    $this->position,
+                    count($this->columns),
+                    count($line)
+                ));
+            }
+
+            $line = array_combine($this->columns, $line);
+        }
+
+        return $line;
```
Comment (greg0ire) [inline src/Source/CsvSourceIterator.php]: You can make this less indented:

```php
    if (!$this->hasHeaders) {
        return $line;
    }
    if ($line === array(null) || $line === null) {
	$this->currentLine = null;

	return;
    }

    if (count($this->columns) !== count($line)) {
	throw new \RuntimeException(sprintf('Invalid column count at line %s. Expected %d; Got %d.',
	    $this->position,
	    count($this->columns),
	    count($line)
	));
    }

    return array_combine($this->columns, $line);
```

File: src/Source/CsvSourceIterator.php (hunk 10 above)
Comment (srascar) [inline src/Source/CsvSourceIterator.php]: This was added according to @rande previous comment

File: src/Source/CsvSourceIterator.php [hunk 13]
```diff
@@ -145,13 +145,77 @@ public function rewind()
...
+    }
+
+    /**
+     * @param array $line
+     *
+     * @return array|void
+     */
+    private function combineColumns($line)
+    {
+        if ($this->hasHeaders) {
```
Comment (greg0ire) [inline src/Source/CsvSourceIterator.php]: You forgot the `!`

**Comment by greg0ire** [comment]
Why does this target master?

**Comment by OskarStark** [comment]
If you want to finish this PR please reopen.

Thank you!

**Pull request closed without merging**